Make a compressed LHCb dataset

mesmith75 opened the pull request:
Saves about a factor 30 in the repository data size for a 2012 dataset.

Fixes #1344

mesmith75 committed:
Make a compressed LHCb dataset

mesmith75 committed:
sort OfflineGangaDiracSplitter

mesmith75 committed:
tidy

mesmith75 committed:
tidy

egede reviewed (commented):
With a whole new dataset object, we will need some tests as well.

mesmith75 committed:
dataset comparison and fix slicing

mesmith75 committed:
add datasets

mesmith75 committed:
tests

mesmith75 committed:
sort extend method

mesmith75 committed:
improve compressed dataset

mesmith75 committed:
fix tests

mesmith75 committed:
Add metadata

mesmith75 committed:
fix slicing

mesmith75 committed:
add dataset configurables

mesmith75 committed:
some things

mesmith75 committed:
Merge branch 'develop' into lhcbDataset

mesmith75 committed:
git branch -d Merge branch 'lhcbDataset' of github.com:ganga-devs/ganga into lhcbDataset

mesmith75 committed:
sort splitting

mesmith75 committed:
fix iteration

mesmith75 committed:
tidy

mesmith75 commented:
Edit: Made it better

So this is about ready. I think this should only really be used with `BKQuery`.

The `LHCbCompressedDataset` object stores a list of `LHCbCompressedFileSet` object. The `FileSet` keeps the list of LFNs with any common prefix taken off. You you get something like this:
```python
[21:36:51]                                                                                                                                                                                                                                                                     
Ganga In [3]: ds = BKQuery('/LHCb/Collision12/Beam4000GeV-VeloClosed-MagDown/Real Data/Reco14/Stripping21/90000000/SEMILEPTONIC.DST').getDataset(True)                                                                                                                         

[21:36:53]                                                                                                                                                                                                                                                                     
Ganga In [4]: ds                                                                                                                                                                                                                                                               
Ganga Out [4]: 
 LHCbCompressedDataset (
   files = [ LHCbCompressedFileSet (
   lfn_prefix = '/lhcb/LHCb/Collision12/SEMILEPTONIC.DST/00041836',
   suffixes = [3666 Entries of type 'str'] 
 )]  ,
   persistency = None,
   treat_as_inputfiles = False,
   depth = 0,
   XMLCatalogueSlice =    LocalFile (
     namePattern = ,
     localDir = ,
     compressed = False
   ) ,
   credential_requirements =    DiracProxy (
     group = lhcb_user,
     encodeDefaultProxyFileName = False,
     dirac_env = None,
     validTime = None
   ) 
 )
```
The indexing and slicing works close to a normal dataset. An index returns a DiracFile, a slice returns a new compressed object:
```python
[15:43:55]                                                                                                                                                                                                                                                                     
Ganga In [3]: ds[0]                                                                                                                                                                                                                                                            
Ganga Out [3]: 
 DiracFile (
   namePattern = ,
   localDir = None,
   locations = []  ,
   compressed = False,
   lfn = /lhcb/LHCb/Collision12/SEMILEPTONIC.DST/00041836/0008/00041836_00086123_1.semileptonic.dst,
   remoteDir = ,
   guid = ,
   defaultSE = ,
   failureReason = ,
   subfiles = []  ,
   credential_requirements =    DiracProxy (
     group = lhcb_user,
     encodeDefaultProxyFileName = False,
     dirac_env = None,
     validTime = None
   ) 
 )
[21:37:17]                                                                                                                                                                                                                                                                     
Ganga In [5]: ds[0:1]                                                                                                                                                                                                                                                          
Ganga Out [5]: 
 LHCbCompressedDataset (
   files = [ LHCbCompressedFileSet (
   lfn_prefix = '/lhcb/LHCb/Collision12/SEMILEPTONIC.DST/00041836/0008/00041836_00086123_1.semileptonic.dst',
   suffixes = []  
 )]  ,
   persistency = None,
   treat_as_inputfiles = False,
   depth = 0,
   XMLCatalogueSlice =    LocalFile (
     namePattern = ,
     localDir = ,
     compressed = False
   ) ,
   credential_requirements =    DiracProxy (
     group = lhcb_user,
     encodeDefaultProxyFileName = False,
     dirac_env = None,
     validTime = None
   ) 
 )
```

The set comparison operators should all work, and allow for comparison between a compressed dataset and a regular `LHCbDataset` by just comparing the list of LFNs of each.  Similarly you can extend a compressed dataset with a regular dataset.

This works for both running jobs on Dirac and locally.

The disk savings are pretty significant. For one magnet polarity of 2012 a job with the full `LHCbDataset` takes 9.2M in the repository. With this it is 333K.

I have also added a function to the GPI to extract file information from the LFN: luminosity, event stats, run number. I have added some functions to `LHCbCompressedDataset` as well to retrieve these for the given dataset and sum them - particularly useful for slicing and when using MC.

mesmith75 committed:
sort subjob metadata

mesmith75 committed:
fix tests

mesmith75 committed:
docstrings

mesmith75 committed:
fix local submission

mesmith75 committed:
add get file information from Dirac

mesmith75 committed:
tidy

egede commented on ganga/GangaLHCb/Lib/LHCbDataset/BKQuery.py:
```diff
@@ -11,7 +11,10 @@
+from GangaCore.Utility.Config import getConfig
 
+configLHCb = getConfig('LHCb')
+defaultCompressed = configLHCb['compressedDataset']
```
Is there any reason for making this configurable?

mesmith75 replied:
I figured it would be nice for the user to not have to select `True` by default. But maybe it does not matter. I think I will make the compressed option the standard for BK query and get rid of the configurable.

egede commented on ganga/GangaLHCb/Lib/LHCbDataset/BKQuery.py:
```diff
@@ -165,41 +168,34 @@ def getDataset(self):
-        #GangaObject.__createNewList(new_files, files, _createDiracLFN)
+        if compressed:
+            import os
+            from GangaLHCb.Lib.LHCbDataset import LHCbCompressedDataset
```
Imports should preferably be at the beginning of files. Taking away the configurable will also remove the if statement.

mesmith75 replied:
I'll move the import but the if needs to be there in case the user chooses to get the full `LHCbDataset`

egede commented on ganga/GangaLHCb/__init__.py:
```diff
@@ -74,10 +74,12 @@ def guessPlatform():
     defaultLHCbDirac = 'prod'
     configLHCb.addOption('LHCbDiracVersion', defaultLHCbDirac, 'set LHCbDirac version')
+
+    configLHCb.addOption('compressedDataset', False, 'Use the compressed dataset with BKQuery by default')
```
Remove these lines?

mesmith75 committed:
remove configurable

mesmith75 committed:
sort PFNs
